File: itmd-521/labs/week-04/assignment-01.py
from pyspark.sql import SparkSession
from pyspark.sql.types import StructType, StructField, StringType, IntegerType

# Create SparkSession
spark = SparkSession.builder \
    .appName("Divvy Trips Analysis") \
    .getOrCreate()

# Define schema
schema = StructType([
    StructField("trip_id", IntegerType(), True),
    StructField("starttime", StringType(), True),
    StructField("stoptime", StringType(), True),
    StructField("bikeid", IntegerType(), True),
    StructField("tripduration", IntegerType(), True),
    StructField("from_station_id", IntegerType(), True),
    StructField("from_station_name", StringType(), True),
    StructField("to_station_id", IntegerType(), True),
    StructField("to_station_name", StringType(), True),
    StructField("usertype", StringType(), True),
    StructField("gender", StringType(), True),
    StructField("birthyear", StringType(), True)
])

# Read CSV with inferred schema
df_inferred = spark.read.csv("Divvy_Trips_2015-Q1.csv", header=True, inferSchema=True)
print("Inferred Schema:")
df_inferred.printSchema()
print("Number of records:", df_inferred.count())

# Read CSV with programmatically attached schema
df_programmatic = spark.read.csv("Divvy_Trips_2015-Q1.csv", header=True, schema=schema)
print("\nProgrammatic Schema:")
df_programmatic.printSchema()
print("Number of records:", df_programmatic.count())

# Read CSV with schema via DDL
df_ddl = spark.read.option("header", "true").csv("Divvy_Trips_2015-Q1.csv")
df_ddl.createOrReplaceTempView("divvy_trips")
df_ddl_with_schema = spark.sql("SELECT * FROM divvy_trips")
print("\nSchema via DDL:")
df_ddl_with_schema.printSchema()
print("Number of records:", df_ddl_with_schema.count())

# Select Gender based on last name and group by station to
selected_gender = df_ddl_with_schema.selectExpr("CASE WHEN substring(gender, 1, 1) >= 'A' AND substring(gender, 1, 1) <= 'K' THEN 'Female' ELSE 'Male' END AS selected_gender", "to_station_name").groupBy("to_station_name").count()
selected_gender.show(10)

# Stop SparkSession
spark.stop()

File: itmd-521/labs/week-04/assignment-01.scala
import org.apache.spark.sql.{SparkSession, DataFrame}
import org.apache.spark.sql.types._

object Assignment01 {
  def main(args: Array[String]): Unit = {

    // Create a SparkSession
    val spark = SparkSession.builder()
      .appName("Assignment01")
      .config("spark.master", "local")
      .getOrCreate()

    // Define schema for CSV file
    val schema = new StructType()
      .add("trip_id", LongType)
      .add("starttime", TimestampType)
      .add("stoptime", TimestampType)
      .add("bikeid", IntegerType)
      .add("tripduration", IntegerType)
      .add("from_station_id", IntegerType)
      .add("from_station_name", StringType)
      .add("to_station_id", IntegerType)
      .add("to_station_name", StringType)
      .add("usertype", StringType)
      .add("gender", StringType)
      .add("birthyear", IntegerType)

    // Read CSV file with inferred schema
    val df1 = readCSV(spark, "Divvy_Trips_2015-Q1.csv")
    println("DataFrame 1 Schema:")
    df1.printSchema()
    println("Count: " + df1.count())

    // Read CSV file with programmatically defined schema
    val df2 = readCSV(spark, "Divvy_Trips_2015-Q1.csv", schema)
    println("DataFrame 2 Schema:")
    df2.printSchema()
    println("Count: " + df2.count())

    // Read CSV file with DDL-defined schema
    val ddlSchema = "trip_id LONG, starttime TIMESTAMP, stoptime TIMESTAMP, bikeid INT, tripduration INT, from_station_id INT, from_station_name STRING, to_station_id INT, to_station_name STRING, usertype STRING, gender STRING, birthyear INT"
    val df3 = readCSV(spark, "Divvy_Trips_2015-Q1.csv", StructType.fromDDL(ddlSchema))

    println("DataFrame 3 Schema:")
    df3.printSchema()
    println("Count: " + df3.count())

    // Stop the SparkSession
    spark.stop()
  }

  // Function to read CSV with inferred, programmatically defined, or DDL-defined schema
  def readCSV(spark: SparkSession, path: String, schema: StructType = null, ddlSchema: String = null): DataFrame = {
    var df: DataFrame = null
    if (ddlSchema != null) {
      df = spark.read.option("header", "true").option("delimiter", ",").schema(ddlSchema).csv(path)
    } else if (schema != null) {
      df = spark.read.option("header", "true").schema(schema).csv(path)
    } else {
      df = spark.read.option("header", "true").csv(path)
    }
    df
  }
}
